packages/techdocs-common/src/stages/publish/googleStorage.ts

```diff
@@ -0,0 +1,222 @@
+        "Since techdocs.publisher.type is set to 'google_gcs' in your app config, " +
+          'pathToKey, projectId and bucketName are required in techdocs.publisher.google ' +
+          'required to authenticate with Google Cloud Storage.',
+      );
+    }
+
+    const storageClient = new Storage({
+      projectId: projectId,
+      keyFilename: pathToKey,
+    });
+
+    // Check if the defined bucket exists. Being able to connect means the configuration is good
+    // and the storage client will work.
+    storageClient
+      .bucket(bucketName)
+      .getMetadata()
+      .then(() => {
+        logger.info(
+          `Successfully connected to the GCS bucket ${bucketName} in the GCP project ${projectId}.`,
+        );
+      })
+      .catch(reason => {
+        logger.error(
+          `Could not retrieve metadata about the GCS bucket ${bucketName} in the GCP project ${projectId}. ` +
+            'Make sure the GCP project and the bucket exists and the access key located at the path ' +
+            "techdocs.publisher.google.pathToKey defined in app config has the role 'Storage Object Creator'. " +
+            'Refer to https://backstage.io/docs/features/techdocs/using-cloud-storage',
+        );
+        throw new Error(`from GCS client library: ${reason.message}`);
+      });
+
+    return new GoogleGCSPublish(storageClient, bucketName, logger);
+  }
+
+  constructor(
+    private readonly storageClient: Storage,
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the GCS bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<{}> {
+    return new Promise((resolve, reject) => {
+      // Path of all files to upload, relative to the root of the source directory
+      // e.g. ['index.html', 'sub-page/index.html', 'assets/images/favicon.png']
+      const allFilesToUpload: Array<string> = [];
+
+      // Iterate on all the files in the directory and its sub-directories
+      walk(directory)
+        .on('data', (item: walk.Item) => {
+          // GCS manages creation of parent directories if they do not exist.
+          // So collecting path of only the files is good enough.
+          if (item.stats.isFile()) {
+            // Remove the absolute path prefix of the source directory
+            const relativeFilePath = item.path.replace(`${directory}/`, '');
+            allFilesToUpload.push(relativeFilePath);
+          }
+        })
+        .on('error', (err: Error, item: walk.Item) => {
+          const errorMessage = `Unable to read file at ${item.path}. Error ${err.message}`;
+          this.logger.error(errorMessage);
+          reject(errorMessage);
+        })
+        .on('end', () => {
```

Comment:
this whole "get all the files in a dir" should really be broken out into a separate function that you can await here as a oneliner